Bring in Packager and make console commands extendable

LukeTowers opened the pull request:
Companion to wintercms/winter#1259 (UpdateManager refactor). That refactor drives module/plugin/theme install and update operations through Composer instead of the legacy marketplace zip downloads, and streams console command output to the backend UI. This PR adds the Storm-side primitives required for that work:

- **Bring in `winter/packager`** (`^0.4.3`) and add a `Winter\Storm\Packager\Composer` helper that proxies to a shared `Winter\Packager\Composer` instance bound to `base_path()`, plus higher-level helpers used by the new UpdateManager (`getWinterPackages()`, `getAvailableUpdates()`, `getPackageInfoByExtension()`, `getWinterPackagesWithVersion()`, etc.) for identifying Winter packages and their available updates from Composer metadata (`vendor/composer/installed.json`).
- **Add the `Winter\Storm\Foundation\Extension\WinterExtension` interface** — a common contract (path / version / identifier / composer package info) for everything installable on top of Winter (modules, plugins, themes) so the UpdateManager and the extension managers in winter#1259 can treat them uniformly. `ModuleServiceProvider` implements it here via the new `HasComposerPackage` trait; the plugin & theme implementations live in winter#1259.
- **Make console commands extendable** — `Winter\Storm\Console\Command` now uses `ExtendableTrait` + `Emitter`, and overrides `run()` to fire a `command.beforeRun` event once output is initialized. This allows external code (e.g. the web-based updater) to intercept a command's output/lifecycle, such as streaming `winter:update` output to the browser. The override replicates Laravel 9's `Illuminate\Console\Command::run()` and invokes the grandparent `Symfony\Component\Console\Command::run()` directly, with a scoped Termwind renderer swap so nested/captured command output renders to the correct output stream.

Merge considerations:

- The `allow-plugins.php-http/discovery: false` config entry is needed because `winter/packager` currently depends on `php-http/discovery` (which satisfies its `psr/http-client-implementation` requirement at resolution time). Downstream projects will face the Composer plugin allow-prompt on update unless packager drops this dependency first (see review discussion).
- `winter/packager` should receive a stable v1.0 release when this is merged.
- `winter/packager` also brings `composer/composer` into the production dependency tree of every Winter install.

jaxwilko committed:
Added event/extension support to base command

jaxwilko committed:
Added support for module service provider instances to be bound to the application container

jaxwilko committed:
Migrated code from winter to storm

jaxwilko committed:
Added local remember function to avoid system cache

jaxwilko committed:
Added base methods for modules to inherit

jaxwilko committed:
Added packager to storm deps

LukeTowers commented on composer.json:
```diff
@@ -49,7 +49,8 @@
         "wikimedia/minify": "~2.2",
-        "winter/laravel-config-writer": "^1.0.1"
+        "winter/laravel-config-writer": "^1.0.1",
+        "winter/packager": "~0.2"
```
When merging we should probably release v1 of Packager.

LukeTowers commented on src/Support/ModuleServiceProvider.php:
```diff
@@ -90,4 +101,14 @@ protected function loadConfigFrom($path, $namespace)
+
+    public function __toString(): string
+    {
+        return $this->getIdentifier();
```
Where is getIdentifier() implemented?

jaxwilko replied:
In the same file now

jaxwilko committed:
Added identifier and path logic to module provider

jaxwilko committed:
Added method for retrieving packages from composer api

jaxwilko commented on src/Console/Command.php:
```diff
@@ -34,6 +49,41 @@ public function __construct()
+        renderUsing($this->output->getOutput());
+
+        try {
+            // Wizardy
```
```suggestion
            // Calling the grandparent run() method, see: https://www.php.net/manual/en/language.oop5.inheritance.php#100005)
```

jaxwilko committed:
Update src/Console/Command.php

jaxwilko commented on src/Console/Command.php:
```diff
@@ -34,6 +49,41 @@ public function __construct()
+        renderUsing($this->output->getOutput());
+
+        try {
+            // Calling the grandparent run() method, see: https://www.php.net/manual/en/language.oop5.inheritance.php#100005)
```
```suggestion
            // Calling the grandparent run() method, see: https://www.php.net/manual/en/language.oop5.inheritance.php#100005
```

jaxwilko committed:
Update src/Console/Command.php

jaxwilko committed:
Manual merge develop

jaxwilko committed:
Added fix for symfony compatibility

jaxwilko committed:
Added helper to composer to allow fetching of installed winter packages with version info

LukeTowers committed:
Merge branch 'develop' into wip/support-uploading-packages-update-manager-changes

LukeTowers commented on src/Console/Command.php:
```diff
@@ -35,6 +50,41 @@ public function __construct()
         if (!empty($this->replaces)) {
             $this->setAliases($this->replaces);
         }
+
+        $this->extendableConstruct();
+    }
+
+    /**
+     * Override the laravel run function to allow us to run callbacks on the command prior to excution.
+     * Run the console command.
+     *
+     * @param  \Symfony\Component\Console\Input\InputInterface  $input
+     * @param  \Symfony\Component\Console\Output\OutputInterface  $output
+     * @return int
+     */
+    public function run(InputInterface $input, OutputInterface $output): int
+    {
+        $this->output = $this->laravel->make(
+            OutputStyle::class, ['input' => $input, 'output' => $output]
+        );
+
+        $this->components = $this->laravel->make(Factory::class, ['output' => $this->output]);
+
+        $this->fireEvent('beforeRun', [$this]);
+
+        $renderer = Termwind::getRenderer();
+        renderUsing($this->output->getOutput());
+
+        try {
+            // Calling the grandparent run() method, see: https://www.php.net/manual/en/language.oop5.inheritance.php#100005
+            return SymfonyCommand::run(
+                $this->input = $input, $this->output
+            );
+        } finally {
+            $this->untrap();
+            // Restore the original termwind renderer
+            renderUsing($renderer);
+        }
```
```suggestion
        $this->output = $this->laravel->make(
            OutputStyle::class,
            ['input' => $input, 'output' => $output]
        );

        $this->components = $this->laravel->make(Factory::class, ['output' => $this->output]);

        $this->fireEvent('beforeRun', [$this]);

        $renderer = Termwind::getRenderer();
        renderUsing($this->output->getOutput());

        try {
            // Calling the grandparent run() method, see: https://www.php.net/manual/en/language.oop5.inheritance.php#100005
            return SymfonyCommand::run(
                $this->input = $input,
                $this->output
            );
        } finally {
            $this->untrap();
            // Restore the original termwind renderer
            renderUsing($renderer);
        }
```

LukeTowers committed:
Update src/Console/Command.php

LukeTowers committed:
Revert change to HandlesCleanup

Applied in L12 PR instead.

LukeTowers committed:
Cleanup

LukeTowers committed:
Fixes for latest version of packager

jaxwilko committed:
Added useful data to return from getAvailableUpdates

jaxwilko committed:
Added fixes for show command

jaxwilko committed:
Added fix to prevent empty data being cached and removed the show command

jaxwilko committed:
Added fixes to command and added helper for updating package version requirements

LukeTowers commented on src/Packager/Commands/InfoCommand.php:
```diff
@@ -0,0 +1,88 @@
+use Winter\Packager\Exceptions\CommandException;
+use Winter\Packager\Exceptions\WorkDirException;
+
+class InfoCommand extends BaseCommand
```
```suggestion
/**
 * @TODO: Relocate to Winter\Packager
 */
class InfoCommand extends BaseCommand
```

LukeTowers committed:
Update src/Packager/Commands/InfoCommand.php

LukeTowers committed:
Remove unnecessary packager overrides

Requires https://github.com/wintercms/packager/pull/6

LukeTowers committed:
Remove unused Search command

LukeTowers committed:
Replace customized RequireCommand with Composer::getLatestSupportedVersion

LukeTowers committed:
Remove customized Update command

Only missing feature from Packager library is the -W (with all dependencies) flag, we aren't using that right now and can add it to Packager directly if we do need it.

LukeTowers commented on composer.json:
```diff
@@ -49,7 +49,8 @@
         "wikimedia/minify": "~2.2",
-        "winter/laravel-config-writer": "^1.0.1"
+        "winter/laravel-config-writer": "^1.0.1",
+        "winter/packager": "~0.2"
```
```suggestion
        "winter/packager": "^0.4"
```

LukeTowers committed:
Update composer.json

LukeTowers committed:
Rename command beforeRun local event to command.beforeRun and documents it

LukeTowers committed:
Merge branch 'wip/support-uploading-packages-update-manager-changes' of github.com:wintercms/storm into wip/support-uploading-packages-update-manager-changes

LukeTowers committed:
Update imports

LukeTowers commented on src/Packager/Composer.php:
```diff
@@ -0,0 +1,274 @@
+
+/**
+ * @class Composer
+ * @method static Collection|DetailedVersionedPackage|DetailedPackage|VersionedPackage|Package|array|null show(?string $mode = 'installed', string $package = null, bool $noDev = false, bool $latest = false)
```
```suggestion
 * @method static Collection|DetailedVersionedPackage|DetailedPackage|VersionedPackage|Package|array|null show(?string $mode = 'installed', string $package = null, bool $noDev = false, bool $latest = false, bool $returnArray = false)
```

LukeTowers committed:
Update src/Packager/Composer.php

LukeTowers committed:
Move listPackages to Bazaar

LukeTowers committed:
Merge branch 'wip/support-uploading-packages-update-manager-changes' of github.com:wintercms/storm into wip/support-uploading-packages-update-manager-changes

LukeTowers committed:
Tidying

LukeTowers committed:
Static analysis fixes

LukeTowers committed:
Tidying

LukeTowers committed:
Improve performance of working with Composer

LukeTowers committed:
Fix getting composer package version

LukeTowers committed:
Fix usage of composer

LukeTowers committed:
Fix PHP Stan

LukeTowers commented on composer.json:
```diff
@@ -99,5 +100,10 @@
     "minimum-stability": "dev",
-    "prefer-stable": true
+    "prefer-stable": true,
+    "config": {
+        "allow-plugins": {
+            "php-http/discovery": false
+        }
+    }
```
Hopefully we can eliminate the need for this.

LukeTowers commented on src/Support/ModuleServiceProvider.php:
```diff
@@ -1,17 +1,27 @@
+use Illuminate\Support\ServiceProvider as ServiceProviderBase;
+use ReflectionClass;
+use Winter\Storm\Foundation\Extension\WinterExtension;
+use Winter\Storm\Packager\Composer;
```
```suggestion
```

LukeTowers committed:
Update src/Support/ModuleServiceProvider.php

LukeTowers committed:
Merge branch 'develop' into wip/support-uploading-packages-update-manager-changes
